Improve: give a simpler connection dialog a go (tabbed)

vadi2 opened the pull request:
<!-- Keep the title short & concise so anyone non-technical can understand it,
     the title appears in PTB changelogs -->
#### Brief overview of PR changes/additions
The connection dialog, looked through the eyes of a first time player, is quite busy and confusing! People will enter their preferred character name and preferred password into the character/password fields without realising that it's for recording _existing_ information.

Experimenting with fixing this by making less info be visible at all times, while not being too far (just a click) away:

## Draft 1
![Screenshot 2022-01-18 12 44 35 AM](https://user-images.githubusercontent.com/110988/149848526-c09120df-a2fc-4a41-a2f2-ec30152df8dd.png)

![Screenshot 2022-01-18 12 44 45 AM](https://user-images.githubusercontent.com/110988/149848527-eae3566b-5125-4480-bffd-d82c02c3537f.png)

## Draft 2
![Screenshot 2022-01-18 10 41 38 AM](https://user-images.githubusercontent.com/110988/149911659-2e13a437-9c8b-4469-9ea4-e196e0777cf8.png)

![Screenshot 2022-01-18 10 41 46 AM](https://user-images.githubusercontent.com/110988/149911676-f2decb2a-a830-4cef-b1a1-ad683f4a3ac1.png)



#### Motivation for adding to Mudlet
Better first-time player experience - the current experience is not good enough.

vadi2 committed:
Give a simpler connection dialog a go

demonnic commented:
I quite like the way this looks and can see the benefit of separating the optional information off to prevent decision overload right off. Plus honestly for most of my profiles I don't fill in any of the optional information myself and like the idea of having it tucked away.

Kebap commented:
Relates to #2962

Eraene commented:
Yep, this looks good to me.

vadi2 committed:
Fix 'secure' button to align

vadi2 committed:
Draft #2

vadi2 commented:
Thanks for mentioning that earlier discussion @Kebap!

I added some ideas I had from there to a draft 2, screenshots and links updated. Is that also an improvement or a step back?

demonnic commented:
Draft 2 looks pretty good.  I like "Connection Info" for the top section, what about "Advanced" or "Options" for the bottom one?

vadi2 commented:
`Options` sits well with me 👍

vadi2 commented on src/ui/connection_profiles.ui:
```diff
@@ -467,324 +468,359 @@
+             </item>
+            </layout>
+           </widget>
+           <widget class="QWidget" name="page_extra">
```
```suggestion
           <widget class="QWidget" name="page_options">
```

vadi2 commented on src/ui/connection_profiles.ui:
```diff
@@ -467,324 +468,359 @@
+             </sizepolicy>
+            </property>
+            <attribute name="label">
+             <string>Extra</string>
```
```suggestion
             <string>Options</string>
```

vadi2 committed:
Extra -> Options

vadi2 commented:
Ready for review -

![image](https://user-images.githubusercontent.com/110988/149999175-24210932-63c3-4bb7-9d46-fc5ae7acbcb2.png)

SlySven commented:
Can I make the observation that the first is less different from the current version than the second and actually that might be more easy as a transitioning step for existing users (the labelling of the sections is the same IIRC).

As such would it not be a reason to go with the first for now - let it mature for a few minor version number releases - before moving to something more like the second. Apart from anything else it also gives us an opportunity to work out how to fit in controls for, apropos some of the stuff in #4214 and #3791:
* `(int) Host::mRetries`
* `(int) Host::mTimeout`
* *probably* also a putative `(int) Host::retryDelay`
* some knob for "*make connection via IPv4 only / IPv6 only / IPv4 & IPv6 (Happy eyeballs)*"

?

vadi2 commented:
🤔 I'm biased in this so I will let others offer an opinion

demonnic commented:
I don't know that we need a transitionary stage/period between the current and draft 2. It's not a huge departure from what we have now, and I feel like changing the dialog twice over the course of a year is adding more work and potential confusion than just doing it the once.

As for the rest, should probably discuss that in those PRs but they seem like candidates for the Options section to me.

SlySven commented:
> As for the rest, should probably discuss that in those PRs but they seem like candidates for the Options section to me.

I haven't even begun to think about code for them yet - though the original concept for parts of it goes back to before we adopted `git` as our SCM - so the progress is a tiny bit glacial. :stuck_out_tongue_winking_eye:

Kebap commented:
Is this Z shaped line somewhat common for collapsible areas?
I found it not too obvious there would be additional data accessible via click, nor where exactly to click.
I would have expected small + or - signs maybe to indicate stuff hidden or shown respectively.

There is quite a lot whitespace now with the hidden character info, etc. 
The connection window height could easily be cut in half or 2/3 the size.
Maybe just enough to fit the Options section once it gets actively shown.

I found no way to open the optional area with keyboard alone and no mouse support either. Did you miss it?

Also the order of objects accessed with tabulator key seems to have been worsened again with this.
Try starting Mudlet and just press TAB a few times. 
In 4.14.1 you start with profile name, then host address, and go down the list. See #3366 for details.
Now the order seems random, starting with "Informational", then "Remove" and "Copy", later arriving at the connection info finally but starting that section with the field "Port"...
Please reconsider.

SlySven commented:
> Is this Z shaped line somewhat common for collapsible areas?

it represents the tabs on folders in a *filing cabinet*, if you can imagine (the ones I never got around to fitting) tabs a few centimetres wide one on each of them in a staggered arrangement on mine:
![IMG0591A](https://user-images.githubusercontent.com/6163092/150972159-01407049-f61b-402c-beb6-724d44659f1c.jpg)

I guess it is a more stylised version of the "Tab Bar" we use for Multi-view on multiple profiles.

Kebap commented:
Oh, how about we use a tab bar then like we do in preferences window? 
"Connection info" tab next to "Options" tab, and not below each other.
Then it is a same UI most players should expect and know how to handle.
We can then even shrink the connection window height a little bit further.

vadi2 commented:
I'll give that design a go!

vadi2 committed:
Merge branch 'development' into try-simpler-connection-dialog

vadi2 committed:
Rebuild connection dialog with a tab

vadi2 committed:
WIP revised dialog with tabs

vadi2 committed:
Layout tweaks

vadi2 commented:
@Kebap it's now available:
![image](https://user-images.githubusercontent.com/110988/161381011-562c1afc-f2eb-49ab-85ee-1445de8f87d5.png)

![image](https://user-images.githubusercontent.com/110988/161381010-680c4694-dc51-4c4a-8fdb-c37744793244.png)

In my opinion it is more intuitive, but far less aesthetic :(

Kebap commented:
I actually like it better, even though you didn't show dark mode 😋 Also there's too much white space until the description area begins. Otherwise, still very solid idea to hide half of the options.

vadi2 commented:
I'm not a fan, it looks really boring now. Let's see what the decision maker patrons think of this!

mpconley commented:
I like freeing up the space, as it will get used later for more helpful components.

vadi2 commented:
Yup, that's the idea behind this - both options free up space, just in two different ways. Will show screenshots soon.

vadi2 committed:
Merge branch 'development' into try-simpler-connection-dialog

vadi2 committed:
Reduce whitespace per feedback

vadi2 commented:
The decision makers have spoken, tabs it is! Not what I hoped but I will obey: 
![image](https://user-images.githubusercontent.com/110988/162695041-765cf829-0b7b-4b92-94b3-f2a796d5e984.png)

Still some days left on the poll but I doubt it'll change - if it does we'll update the PTB. Meanwhile, finalized version the dialog:

![image](https://user-images.githubusercontent.com/110988/162695135-8e696dce-3c7b-443d-8d3f-74117045e8b1.png)

![image](https://user-images.githubusercontent.com/110988/162695140-28817734-704c-4b48-b2bf-db5e2262974e.png)

SlySven commented:
Bother, I thought I had left a comment about some internal breakages in the `.ui` file but somehow they never got sent to here:
* You have used a `QFormLayout` for one of the two tabs but we have had breakages with that in the past because the styling is style-factory dependent - remember #5352 ? It took me a while to find out that the behaviour is different on MacOs! https://github.com/Mudlet/Mudlet/pull/5372#issuecomment-913047788
* There is a text label that says "Website:"  but it doesn't show currently because it is spread over both rows in the "Informational" area and is hidden behind the `QTextBrowser` widget that it (erroneously) shares the same space with.
* The MUD icons are laid out in a `QListWidget` which is really confusingly called `profiles_tree_widget` at the moment!

Also I think you might need a `QSpacer` set up for vertical operation at the bottom of the second tab to ensure everything is pushed to the top like it is on the first tab!

Ah, also, the current index property for the tab widget is set to `1` currently which means that it will be showing the "Options" tab rather that the "Connection info" one when the dialogue is initially opened...

vadi2 committed:
Change layout to grid

vadi2 committed:
Show connection info by default

vadi2 commented:
Changed the form layout to a grid layout, and I've fixed the dialog to show the connection area first, thanks!

Other feedback was not applied:

- the website label shows correctly, see screenshots above
- `profiles_tree_widget` was called this before the PR and this PR does not touch that part of the the dialog, I'm not going to leak scope to change that as well. That feedback is recorded as https://github.com/Mudlet/Mudlet/issues/6039
- adding the spacer just adds whitespace and removes space from the informational area - that is not a good idea

SlySven commented:
>* the website label shows correctly, see screenshots above

You are incorrect I am afraid: if you look more carefully in the QtDesigner you might be able to detect the following widget which I have simulated and drawn-in in red - the reason it is not visible is because it is drawn over both rows in the layout and as the bottom row is much larger than the top the (vertically centred) text is hidden behind the `QTextBrowser` widget that shows the "Description"/"User notes":
![Borked_informational_area](https://user-images.githubusercontent.com/6163092/163825020-074b2f3d-da33-4431-a419-b10700ebd920.png)

SlySven commented:
FYI: The line that causes this is on old 823, new 828, for the `QLabel` named `label_7` I *think* the fix would be:
```diff
- <item row="0" column="0" rowspan="2">
+ <item row="0" column="0">
 ```
but I am not entirely sure that the alignment looks right when this is done.

vadi2 commented:
This has nothing at all to do with this improvement - it's been like this for years.

vadi2 committed:
Simplify string

Kebap reviewed (approved):
Let's go
